tools/concrete-protocol/src/concrete-protocol.proto

```diff
@@ -0,0 +1,295 @@
+
+message ServerKeyset {
+    repeated LweBootstrapKey lweBootstrapKeys = 1;
+    repeated LweKeyswitchKey lweKeyswitchKeys = 2;
+    repeated PackingKeyswitchKey packingKeyswitchKeys = 3;
+}
+
+message ClientKeyset {
+    repeated LweSecretKey lweSecretKeys = 1;
+}
+
+message Keyset {
+    ServerKeyset server = 1;
+    ClientKeyset client = 2;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////// Encoding
+
+message IntegerCiphertextEncodingInfo {
+    uint32 width = 1;
+    bool isSigned = 2;
+
+    message NativeMode {}
+    
+    message ChunkedMode {
+        uint32 size = 1;
+        uint32 width = 2;
+    }
+
+    message CrtMode {
+        repeated uint32 moduli = 1;
+    }
+
+    oneof mode {
+        NativeMode native = 3;
+        ChunkedMode chunked = 4;
+        CrtMode crt = 5;
+    }
+}
+
+message BooleanCiphertextEncodingInfo {}
+
+message PlaintextEncodingInfo {}
+
+message IndexEncodingInfo {}
+
+message EncodingInfo {
+    optional Shape shape = 1;
+    oneof encoding {
+        IntegerCiphertextEncodingInfo integerCiphertext = 2;
+        BooleanCiphertextEncodingInfo booleanCiphertext = 3;
+        PlaintextEncodingInfo plaintext = 4;
+        IndexEncodingInfo index = 5;
+    }
+}
+
+message CircuitEncodingInfo {
+    repeated EncodingInfo inputs = 1;
+    repeated EncodingInfo outputs = 2;
+    string name = 3;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////// Encryption
+
+message LweCiphertextEncryptionInfo {
+    uint32 keyId = 1;
+    double variance = 2;
+    uint32 integerPrecision = 3;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////// Gates
+
+message GateInfo {
+    optional Shape shape = 1;
+    oneof kind {
+        LweCiphertextGateInfo lweCiphertext = 2;
+        PlaintextGateInfo plaintext = 3;
+        IndexGateInfo index = 4;
+    }
+}
+
+message LweCiphertextGateInfo {
+    LweCiphertextEncryptionInfo encryption = 1;
+    oneof encoding {
+        IntegerCiphertextEncodingInfo integer = 2;
+        BooleanCiphertextEncodingInfo boolean = 3;
+    }
+}
+
+message PlaintextGateInfo {
+    uint32 integerPrecision = 1;
+    bool isSigned = 2;
+}
+
+message IndexGateInfo {
+    uint32 integerPrecision = 1;
+    bool isSigned = 2;
+}
+
+/////////////////////////////////////////////////////////////////////////////// Circuit Description
+
+message CircuitInfo {
+    repeated GateInfo inputs = 1;
+    repeated GateInfo outputs = 2;
+    string name = 3;
+}
+
+message ProgramInfo {
+    KeysetInfo keyset = 1;
+    repeated CircuitInfo circuits = 2;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////// Values
+
+message Value {
+    ValueInfo info = 1;
+    bytes data = 2;
+}
+
+message ValueInfo {
+    optional Shape shape = 1;
+    oneof kind {
+        LweCiphertextValueInfo lweCiphertext = 2;
+        PlaintextValueInfo plaintext = 3;
```

Comment:
I'm wondering why it is important to differentiate plaintext and index? Is it if the client and server doesn't have the same architecture?

Reply:
The first motivation is to differentiate two completely different kind of inputs. The second motivation is that plaintexts must have the same integer precision as every other plaintexts and ciphertexts in the circuit, whereas the index must not.